File: mobile/src/main/java/com/windscribe/mobile/welcome/fragment/WelcomeFragment.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.mobile.welcome.fragment;

import static androidx.fragment.app.FragmentStatePagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateInterpolator;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.bumptech.glide.Glide;
import com.google.android.material.tabs.TabLayout;
import com.windscribe.mobile.R;
import com.windscribe.mobile.fragments.FeatureFragments;
import com.windscribe.mobile.fragments.FeaturePageTransformer;

import java.util.Timer;
import java.util.TimerTask;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class WelcomeFragment extends Fragment implements ViewPager.OnPageChangeListener {

    @BindView(R.id.logo)
    ImageView logo;

    @BindView(R.id.feature_pager)
    ViewPager mViewPager;

    @BindView(R.id.featureTabDots)
    TabLayout tabLayout;

    private FragmentCallback fragmentCallback;

    private Runnable mAutoPagingRunnable;

    private PagerAdapter mPagerAdapter;

    private int mScrollState = 0;

    private boolean mSlideLeft = true;

    private Timer pagerTimer;

    public WelcomeFragment() {

    }

    @Override
    public void onAttach(@NonNull Context context) {
        if (getActivity() instanceof FragmentCallback) {
            fragmentCallback = (FragmentCallback) getActivity();
        }
        super.onAttach(context);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_welcome, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setGif();
        setPagerAdapter();
        setUpAutoPaging();
    }

    @Override
    public void onDestroyView() {
        stopPagerSchedule();
        super.onDestroyView();
    }

    @Override
    public void onPageScrollStateChanged(int state) {
        mScrollState = state;
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
    }

    @Override
    public void onPageSelected(int position) {
    }

    public void setPagerAdapter() {
        mPagerAdapter = new FragmentPagerAdapter(getChildFragmentManager(), BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
            @Override
            public int getCount() {
                return 4;
            }

            @NonNull
            @Override
            public Fragment getItem(int position) {
                return FeatureFragments.newInstance(position);
            }
        };
        tabLayout.setupWithViewPager(mViewPager, true);
        mViewPager.addOnPageChangeListener(this);
        mViewPager.setPageTransformer(false, new FeaturePageTransformer());
        mViewPager.setAdapter(mPagerAdapter);
    }

    public void setUpAutoPaging() {
        mAutoPagingRunnable = () -> {
            if (!mViewPager.isFakeDragging() && mScrollState == 0) {
                mViewPager.beginFakeDrag();
                ValueAnimator valueAnimator = ValueAnimator.ofFloat(0, 1);
                valueAnimator.setInterpolator(new AccelerateInterpolator(1.5f));
                valueAnimator.setDuration(2000);
                valueAnimator.addListener(new Animator.AnimatorListener() {
                    @Override
                    public void onAnimationCancel(Animator animator) {
                        if (mViewPager == null) {
                            return;
                        }
                        mViewPager.endFakeDrag();
                    }

                    @Override
                    public void onAnimationEnd(Animator animator) {
                        try {
                            if (mViewPager == null | mPagerAdapter == null) {
                                return;
                            }
                            if (mViewPager.isFakeDragging()) {
                                mViewPager.endFakeDrag();
                            }
                            if (mViewPager.getCurrentItem() == mPagerAdapter.getCount() - 1) {
                                mSlideLeft = false;
                            } else if (mViewPager.getCurrentItem() == 0) {
                                mSlideLeft = true;
                            }
                        } catch (Exception ignored) {
                        }
                    }

                    @Override
                    public void onAnimationRepeat(Animator animator) {

                    }

                    @Override
                    public void onAnimationStart(Animator animator) {
                        mViewPager.beginFakeDrag();
                    }
                });
                valueAnimator.addUpdateListener(valueAnimator1 -> {
                    if (mViewPager == null | mPagerAdapter == null) {
                        return;
                    }
                    try {
                        if (mViewPager.isFakeDragging() && mPagerAdapter.getCount() > 0) {
                            if (mSlideLeft) {
                                mViewPager.fakeDragBy(
                                        -valueAnimator1.getAnimatedFraction() * mViewPager.getWidth() / 2);
                            } else {
                                mViewPager
                                        .fakeDragBy(valueAnimator1.getAnimatedFraction() * mViewPager.getWidth() / 2);
                            }
                        }
                    } catch (Exception ignored) {
                    }
                });
                valueAnimator.start();
            }
        };
        startPagerSchedule();
    }

    public void startPagerSchedule() {
        Handler handler = new Handler();
        pagerTimer = new Timer();
        pagerTimer.schedule(new TimerTask() {
            @Override
            public void run() {
                handler.post(mAutoPagingRunnable);
            }
        }, 3000, 3000);
    }

    public void stopPagerSchedule() {
        pagerTimer.cancel();
        pagerTimer = null;
    }

    @SuppressLint("NonConstantResourceId")
    @OnClick(R.id.get_started_button)
    void onGetStartedButtonClick() {
        fragmentCallback.onContinueWithOutAccountClick();
    }

    @SuppressLint("NonConstantResourceId")
    @OnClick(R.id.loginButton)
    void onLoginButtonClick() {
        fragmentCallback.onLoginClick();
    }

    private void setGif() {
        Glide.with(this).load(R.raw.wsbadge).into(logo);
    }
}

File: mobile/src/main/java/com/windscribe/mobile/welcome/fragment/WelcomeFragment.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */
package com.windscribe.mobile.welcome.fragment

import android.animation.Animator
import android.animation.ValueAnimator
import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateInterpolator
import android.widget.ImageView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import com.bumptech.glide.Glide
import com.google.android.material.tabs.TabLayout
import com.windscribe.mobile.R
import com.windscribe.mobile.fragments.FeatureFragments
import com.windscribe.mobile.fragments.FeaturePageTransformer
import java.util.*

class WelcomeFragment : Fragment(), OnPageChangeListener {
    @BindView(R.id.logo)
    lateinit var logo: ImageView

    @BindView(R.id.feature_pager)
    lateinit var mViewPager: ViewPager

    @BindView(R.id.featureTabDots)
    lateinit var tabLayout: TabLayout

    private var fragmentCallback: FragmentCallback? = null
    private var autoPagingRunnable: Runnable? = null
    private var pagerAdapter: PagerAdapter? = null
    private var scrollState = 0
    private var slideLeft = true
    private var pagerTimer: Timer? = null

    override fun onAttach(context: Context) {
        if (activity is FragmentCallback) {
            fragmentCallback = activity as FragmentCallback?
        }
        super.onAttach(context)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(R.layout.fragment_welcome, container, false)
        ButterKnife.bind(this, view)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setGif()
        setPagerAdapter()
        setUpAutoPaging()
    }

    override fun onDestroyView() {
        stopPagerSchedule()
        super.onDestroyView()
    }

    override fun onPageScrollStateChanged(state: Int) {
        scrollState = state
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
    override fun onPageSelected(position: Int) {}
    private fun setPagerAdapter() {
        pagerAdapter = object :
            FragmentPagerAdapter(childFragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
            override fun getCount(): Int {
                return 4
            }

            override fun getItem(position: Int): Fragment {
                return FeatureFragments.newInstance(position)
            }
        }
        tabLayout.setupWithViewPager(mViewPager, true)
        mViewPager.addOnPageChangeListener(this)
        mViewPager.setPageTransformer(false, FeaturePageTransformer())
        mViewPager.adapter = pagerAdapter
    }

    private fun setUpAutoPaging() {
        autoPagingRunnable = Runnable {
            if (!mViewPager.isFakeDragging && scrollState == 0) {
                mViewPager.beginFakeDrag()
                val valueAnimator = ValueAnimator.ofFloat(0f, 1f)
                valueAnimator.interpolator = AccelerateInterpolator(1.5f)
                valueAnimator.duration = 2000
                valueAnimator.addListener(object : Animator.AnimatorListener {
                    override fun onAnimationCancel(animator: Animator) {
                        mViewPager.endFakeDrag()
                    }

                    override fun onAnimationEnd(animator: Animator) {
                        try {
                            if (mViewPager.isFakeDragging) {
                                mViewPager.endFakeDrag()
                            }
                            if (mViewPager.currentItem == (pagerAdapter?.count ?: 0) - 1) {
                                slideLeft = false
                                mViewPager.currentItem = 0
                            } else if (mViewPager.currentItem == 0) {
                                slideLeft = true
                            }
                        } catch (ignored: Exception) {
                        }
                    }

                    override fun onAnimationRepeat(animator: Animator) {}
                    override fun onAnimationStart(animator: Animator) {
                        mViewPager.beginFakeDrag()
                    }
                })
                valueAnimator.addUpdateListener { valueAnimator1: ValueAnimator ->
                    if (pagerAdapter == null) {
                        return@addUpdateListener
                    }
                    try {
                        if (mViewPager.isFakeDragging && (pagerAdapter?.count ?: 0) > 0) {
                            if (slideLeft) {
                                mViewPager.fakeDragBy(
                                    -valueAnimator1.animatedFraction * mViewPager.width / 2
                                )
                            }
                        }
                    } catch (ignored: Exception) {
                    }
                }
                valueAnimator.start()
            }
        }
        startPagerSchedule()
    }

    private fun startPagerSchedule() {
        val handler = Handler()
        pagerTimer = Timer()
        pagerTimer?.schedule(object : TimerTask() {
            override fun run() {
                autoPagingRunnable?.let { handler.post(it) }
            }
        }, 3000, 3000)
    }

    private fun stopPagerSchedule() {
        pagerTimer?.cancel()
        pagerTimer = null
    }

    @OnClick(R.id.get_started_button)
    fun onGetStartedButtonClick() {
        fragmentCallback?.onContinueWithOutAccountClick()
    }

    @OnClick(R.id.loginButton)
    fun onLoginButtonClick() {
        fragmentCallback?.onLoginClick()
    }

    @OnClick(R.id.emergencyConnectButton)
    fun onEmergencyButtonClick() {
        fragmentCallback?.onEmergencyClick()
    }

    private fun setGif() {
        Glide.with(this).load(R.raw.wsbadge).into(logo)
    }
}